proposals/0010-validating-resource-container-elements.md

```diff
@@ -0,0 +1,142 @@
+* Proposal: [0010](0010-validating-resource-container-elements.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+*During the review process, add the following fields as needed:*
+
+* PRs: [#NNNN](https://github.com/microsoft/DirectXShaderCompiler/pull/NNNN)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, and float). Structs that contain fields of these basic
+types (where all fields in the struct have the same type) may also be RETs. 
+Structs that either have structs as fields or arrays of structs as fields may also be
+allowed, as long as everything can fit in 4 32-bit quantities. Additionally, resource 
+types are not allowed within an RET, even if the underlying resource type has a 
+primitive RET.
+
+RETs for `RawBuffer` variants are much less constrained, the only rule is that the RET
+may not be an incomplete type (a handle type or a resource type).

```

Comment:
Can HLSL authors interact with handle types directly?

Reply:
It should technically be possible.
I wrote a test in `clang\test\SemaHLSL\resource_binding_attr_error.hlsl` where `__hlsl_resource_t`, the handle type, was directly declared, so it is spellable.
But I don't think we intend users to interact with them.

Reply:
Technically the `__hlsl_resource_t` type is spellable, but there is not much one can do with it. It needs to be decorated with resource type attributes to signify what kind of resource it is representing, and it will never get automatic binding unless it is embedded in a struct that looks like a resource class (has the handle as a first field named "h").